compiler/rustc_const_eval/src/interpret/intern.rs

```diff
@@ -332,122 +99,105 @@ pub fn intern_const_alloc_recursive<
+            continue;
+        }
+        just_interned.insert(alloc_id);
+        intern_shallow(ecx, alloc_id, mutability, |ecx, prov| {
+            let alloc_id = prov.alloc_id();
+            if intern_kind != InternKind::Promoted
+                && inner_mutability == Mutability::Not
+                && !prov.immutable()
+            {
+                if ecx.tcx.try_get_global_alloc(alloc_id).is_some()
+                    && !just_interned.contains(&alloc_id)
+                {
+                    // This is a pointer to some memory from another constant. We encounter mutable
+                    // pointers to such memory since we do not always track immutability through
+                    // these "global" pointers. Allowing them is harmless; the point of these checks
+                    // during interning is to justify why we intern the *new* allocations immutably,
+                    // so we can completely ignore existing allocations. We also don't need to add
+                    // this to the todo list, since after all it is already interned.
+                    return;
                 }
+                // Found a mutable pointer inside a const where inner allocations should be
+                // immutable. We exclude promoteds from this, since things like `&mut []` and
+                // `&None::<Cell<i32>>` lead to promotion that can produce mutable pointers. We rely
+                // on the promotion analysis not screwing up to ensure that it is sound to intern
+                // promoteds as immutable.
+                found_bad_mutable_pointer = true;
             }
-        } else if ecx.memory.dead_alloc_map.contains_key(&alloc_id) {
-            // Codegen does not like dangling pointers, and generally `tcx` assumes that
-            // all allocations referenced anywhere actually exist. So, make sure we error here.
-            let reported = ecx.tcx.dcx().emit_err(DanglingPtrInFinal { span: ecx.tcx.span });
-            return Err(reported);
-        } else if ecx.tcx.try_get_global_alloc(alloc_id).is_none() {
-            // We have hit an `AllocId` that is neither in local or global memory and isn't
-            // marked as dangling by local memory. That should be impossible.
-            span_bug!(ecx.tcx.span, "encountered unknown alloc id {:?}", alloc_id);
-        }
+            // It is tempting to intern as immutable if `prov.immutable()`. However, there
+            // might be multiple pointers to the same allocation, and if *at least one* of
+            // them is mutable, the allocation must be interned mutably. We will intern the
+            // allocation when we encounter the first pointer. Therefore we always intern
+            // with `inner_mutability`, and furthermore we ensured above that if that is
+            // "immutable", then there are *no* mutable pointers anywhere in the newly
+            // interned memory.
```

Comment:
This explanation makes it sound like we do some sort of "find all pointers, check if any are mutable" analysis, even if the end of the comment explains that this is not so. Maybe start a sentence with `Instead, ...` so it's clear that we don't do what we earlier said we *must* do

Reply:
I am confused by your comment... we do "find all pointers, check if any are mutable". We error if that check finds anything.

I have reworded the comment, does that help?

Reply:
oh... that "must" was descriptive, not prescriptive 🤦